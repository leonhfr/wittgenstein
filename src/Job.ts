// Internal.
import { Zone } from './Zone';
import * as Types from './types';

// Code.
export class Job {
  readonly page: number;
  readonly zone: Zone;

  static create(input: unknown): Job | Error {
    if (isJob(input)) {
      const { page, zone } = input;
      const maybeZone = Zone.create(zone);

      if (maybeZone instanceof Error) {
        return new Error(`[Zone Validation] ${maybeZone.message}`);
      }

      return new Job({ page, zone: maybeZone });
    }

    const errMsg = isSafeJob(input).errMsg;
    return new Error(`ValidationError: ${errMsg}`);
  }

  private constructor(input: CreateJobInput) {
    this.page = input.page;
    this.zone = input.zone;
  }
}

// Validation.

export const JOB_PROPS = ['page', 'zone'];

export const isJob = (input: unknown): input is CreateJobInput => {
  return isSafeJob(input).isSafe;
};

export const isSafeJob = (input: unknown): Types.IsSafe => {
  if (typeof input !== 'object') {
    return {
      isSafe: false,
      errMsg: `Expected type of input to be object, got ${typeof input}.`,
    };
  }

  if (input === null) {
    return {
      isSafe: false,
      errMsg: `Expected input not to be null.`,
    };
  }

  for (const key of JOB_PROPS) {
    if (!input.hasOwnProperty(key)) {
      return {
        isSafe: false,
        errMsg: `Expected all input properties to be set, missing ${key}.`,
      };
    }
  }

  const { page } = input as {
    page: unknown;
  };

  const numbers = { page };

  // Numbers

  for (const [key, value] of Object.entries(numbers)) {
    if (typeof value !== 'number') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be number, got ${typeof value}.`,
      };
    }
  }

  return {
    isSafe: true,
    errMsg: '',
  };
};

// Interface

export interface CreateJobInput {
  readonly zone: Zone;
  readonly page: number;
}
