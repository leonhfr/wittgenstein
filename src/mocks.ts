// Internal.
import { CreateAreaInput } from './Area';
import { CreatePhotoInput } from './Photo';

// Code.
export const createMockAreaInput = (area?: any): CreateAreaInput => {
  const _area = area || {};
  return {
    id: _area.id || '8ea8f412-4ee8-4f1d-9228-4c133b8fbd37',
    name: _area.name || 'Sant Marti',
    file: _area.file || 'sant-marti',
    lastScheduledAt: _area.lastScheduledAt || 0,
    refreshRate: _area.refreshRate || 86400,
    enabled: _area.enabled || true,
    zonesComputed: _area.zonesComputed || false,
  };
};

export const createMockPhotoInput = (photo?: any): CreatePhotoInput => {
  const _photo = photo || {};
  return {
    id: _photo.id || '32457360587',
    owner: _photo.owner || '41710899@N08',
    secret: _photo.secret || '69b27ceb18',
    server: _photo.server || '4898',
    farm: _photo.farm || 5,
    title: _photo.title || 'Arc de Triomf',
    description:
      _photo.description ||
      'The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.',
    ownername: _photo.ownername || 'jdf_92',
    views: _photo.views || 16,
    tags: _photo.tags || [
      'spain',
      'barcelona',
      'arcdetriomf',
      '1888barcelonaworldfair',
      'neomudéjar',
      '1888barcelonauniversalexposition',
      'catelonia',
    ],
    latitude: _photo.latitude || 41.390791,
    longitude: _photo.longitude || 2.18114,
    context: _photo.context || 0,
  };
};
