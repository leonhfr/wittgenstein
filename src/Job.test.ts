// Internal.
import { Job, JOB_PROPS, isJob } from './Job';
import * as Mocks from './mocks';

// Code.
describe('Job', () => {
  describe('Class', () => {
    it('should be defined', () => {
      expect(Job).toBeDefined();
    });

    describe('create', () => {
      it('should fail to create a new instance - invalid input', () => {
        expect(Job.create(null)).toMatchSnapshot();
      });
      it('should create a new instance', () => {
        expect(Job.create(Mocks.createMockJobInput())).toMatchSnapshot();
      });
      it('should handle an invalid input - zone', () => {
        expect(
          Job.create({ ...Mocks.createMockJobInput(), zone: false })
        ).toMatchSnapshot();
      });
    });
  });

  describe('Validation', () => {
    it('should be defined', () => {
      expect(JOB_PROPS).toBeDefined();
    });
    it('should match the expected value', () => {
      expect(JOB_PROPS).toMatchSnapshot();
    });
    it('should be up to date', () => {
      expect(JOB_PROPS).toEqual(
        Object.keys(Job.create(Mocks.createMockJobInput()))
      );
    });
  });

  describe('isJob', () => {
    it('should be defined', () => {
      expect(isJob).toBeDefined();
    });
    it('should handle an invalid input - wrong type', () => {
      expect(isJob('__FAIL__')).toBe(false);
    });
    it('should handle an invalid input - null type', () => {
      expect(isJob(null)).toBe(false);
    });
    it('should handle an invalid input - missing props', () => {
      expect(isJob({})).toBe(false);
    });
    it('should handle an invalid input - numbers', () => {
      expect(
        isJob({
          ...Mocks.createMockJobInput(),
          page: false,
        })
      ).toBe(false);
    });
  });
});
