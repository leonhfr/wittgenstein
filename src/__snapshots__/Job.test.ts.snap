// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Class create should create a new instance 1`] = `
Job {
  "page": 1,
  "zone": Zone {
    "area": Array [
      "8ea8f412-4ee8-4f1d-9228-4c133b8fbd37",
    ],
    "bbox": Array [
      2.18,
      41.4,
      2.19,
      41.5,
    ],
    "id": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
    "zone": Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              2.18,
              41.4,
            ],
            Array [
              2.19,
              41.5,
            ],
            Array [
              2.19,
              41.4,
            ],
            Array [
              2.18,
              41.4,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "properties": Object {},
      "type": "Feature",
    },
  },
}
`;

exports[`Job Class create should fail to create a new instance - invalid input 1`] = `[Error: ValidationError: Expected input not to be null.]`;

exports[`Job Class create should handle an invalid input - zone 1`] = `[Error: [Zone Validation] ValidationError: Expected type of input to be object, got boolean.]`;

exports[`Job Validation should match the expected value 1`] = `
Array [
  "page",
  "zone",
]
`;
